### PR TITLE
Detect light client attacks and report them to both the chain and the RPC node

### PR DESCRIPTION
**Pull request by romac**
Closes: #2772

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Review by ancazamfir — CHANGES_REQUESTED** [review]
Great work @romac!!! Did a fist pass in parallel with the `tendermint-rs` changes.

File: crates/relayer/src/light_client/tendermint.rs [hunk 1]
```diff
@@ -137,47 +128,83 @@ impl super::LightClient<CosmosSdkChain> for LightClient {
...
+            .map_err(|e| Error::light_client_io(self.chain_id.to_string(), e))?;
+
+        let target_block: LightBlock = LightBlock {
+            signed_header: update_header.signed_header.clone(),
+            validators: update_header.validator_set.clone(),
+            next_validators,
+            provider: self.peer_id,
+        };
+
+        let now = latest_chain_block.time(); // FIXME: Is that right?
```
[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer/src/light_client/tendermint.rs]: 9b9ffce8cf5372c3d66fb6f5271d65c8c55c92ea

File: crates/relayer/src/light_client/tendermint.rs [hunk 2]
```diff
@@ -137,47 +128,83 @@ impl super::LightClient<CosmosSdkChain> for LightClient {
...
+                        client_id: update.client_id().clone(),
+                        header1: update_header.clone(),
+                        header2: TmHeader {
+                            signed_header: challenging_block.signed_header,
+                            validator_set: challenging_block.validators,
+                            trusted_height: update_header.trusted_height,
+                            trusted_validator_set: trusted_block.validators,
+                        },
+                    }),
+                    supporting_headers: vec![], // FIXME: What do we put here?
```
[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer/src/light_client/tendermint.rs:199]: Won't the trace typically only contain the target and trusted headers? Do we really need to supply them as supporting headers in that case?

File: crates/relayer/src/foreign_client.rs [hunk 3]
```diff
@@ -1747,14 +1737,17 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
...
                     MisbehaviourResults::CannotExecute
                 }
+
                 ForeignClientErrorDetail::ExpiredOrFrozen(_) => {
                     error!("cannot check misbehavior on frozen or expired client",);
 
                     MisbehaviourResults::CannotExecute
                 }
 
+                // FIXME: This is fishy
```
Comment (romac) [inline crates/relayer/src/foreign_client.rs:1747]: FIXME: This is fishy

File: crates/relayer/src/foreign_client.rs [hunk 4]
```diff
@@ -1747,14 +1737,17 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
...
                 _ if update_event.is_some() => MisbehaviourResults::CannotExecute,
 
+                // FIXME: This is fishy
```
Comment (romac) [inline crates/relayer/src/foreign_client.rs:1750]: FIXME: This is fishy

File: crates/relayer/src/foreign_client.rs [hunk 5]
```diff
@@ -1443,6 +1443,7 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
         &self,
         mut update: Option<&UpdateClient>,
     ) -> Result<Option<MisbehaviourEvidence>, ForeignClientError> {
+        // FIXME(romac): Why do we need this, and shouldn't we wait somewhere else up the call stack?
```
[the remark this answers is not in the document]
Comment (ancazamfir) [reply crates/relayer/src/foreign_client.rs:1446]: I am not sure. Let's open a separate issue and investigate there.

File: crates/relayer/src/light_client/tendermint.rs [hunk 6]
```diff
@@ -137,47 +139,93 @@ impl super::LightClient<CosmosSdkChain> for LightClient {
...
+            .map_err(|e| Error::light_client_io(self.chain_id.to_string(), e))?;
+
+        let target_block: LightBlock = LightBlock {
+            signed_header: update_header.signed_header.clone(),
+            validators: update_header.validator_set.clone(),
+            next_validators,
+            provider: self.peer_id,
+        };
+
+        let trusted_block = self.fetch(update_header.trusted_height)?; // TODO: Check validator sets match
```
[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer/src/light_client/tendermint.rs:165]: Right, I hadn't thought of that. Yeah let's go with an error for now.

[the remark this answers is not in the document]
Comment (romac) [reply crates/relayer/src/light_client/tendermint.rs]: [`5f253fe` (#3224)](https://github.com/informalsystems/hermes/pull/3224/commits/5f253fef11815fa30838d01b5dbc67e2689012cd)

**Review by ancazamfir — APPROVED** [review]
Amazing job @romac!

File: crates/relayer/src/foreign_client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ancazamfir) [reply crates/relayer/src/foreign_client.rs:1481]: see #3351

File: crates/relayer/src/foreign_client.rs (hunk 4 above)
Comment (ancazamfir) [reply crates/relayer/src/foreign_client.rs:1785]: see #3351

**Pull request merged**